### ISSUE TITLE
When an initial color is set in flat-mode, it cannot be re-chosen

### ISSUE THREAD

**Issue by mydea** [bgrins/spectrum#264]
When creating a flat color picker and setting an initial value with ``color: rgb(255,0,0)`` it is impossible to re-choose this value.

This jsfiddle shows the problem clearly:

http://jsfiddle.net/qfo2ykt0/5/

The color is selected correctly at the beginning, and it appears to work if I choose another color and then the initial color again (it is marked correctly). However, the actual value is not updated, and no change event is fired.

It seems to be coupled to the flat color picker, if you set ```flat: false``` it works as it should!
Note that it makes no difference if I set the initial value with ``color: rgb(255,0,0)`` or on the input field itself.

**Comment by pettarin** [comment]
Just a note: if you set `flat: false` (or you omit `flat`) it _does_ update the color box, but the `onchange` event is still _not_ emitted.

**Comment by pettarin** [comment]
I guess the issue stems from `hasChanged` ( https://github.com/bgrins/spectrum/blob/master/spectrum.js#L850 ): if you start with (say) black, then select (say) red, then select again black, `hasChanged` is `false` there.

For my needs, I simply removed the `&& hasChanged)` part of the `if ` at https://github.com/bgrins/spectrum/blob/master/spectrum.js#L862 , but probably you might want to add an actual sentinel/flag, that becomes true (and stays true) when a color, different than the original one, has been chosen.

**Comment by joeax** [comment]
I had the same issue as pettarin. I considered the same fix but I realized this just covers up the problem. The correct fix would be to update colorOnShow inside of paletteElementClick() right after the call to updateOriginalInput() ([near line 470](https://github.com/bgrins/spectrum/blob/master/spectrum.js#L470)). To me, since the change event is fired the updated color is effectively "committed".

```javascript
updateOriginalInput(true);
colorOnShow = $(e.target).closest(".sp-thumb-el").data("color"); // I added this fix here
```

Another solution is to add a `paletteSelectionFiresChange` option to enable turning off firing change after the palette color is clicked.

**Comment by jtallieu** [comment]
Any update on when this might get released?

**Comment by kotmatpockuh** [comment]
@jtallieu 

https://github.com/kotmatpockuh/spectrum/commit/324935265104a0ecccb7a1c28d77b0e6ce197abf

maybe my fork is usefull...

it has method 'changeInitialColor'

**Comment by Netrilix** [comment]
@kotmatpockuh I wish I had seen that before I dug into it myself in #326. I looks like we did the same thing, except I did it inside `updateOriginalInput()` and you did it separately.

**Comment by lulessa** [comment]
I'm also having trouble with this issue, even without using flat mode. See #326

@kotmatpockuh @Netrilix 
How about one of you makes a pull request with your fork or just a branch with your solution?
CI will run tests on it, so you can tell if it's breaking something else, and @bgrins can just merge it if he decides it makes sense

**Comment by lulessa** [comment]
There's a branch in the repo with passing tests and code by @bgrins to fix this issue:
https://github.com/bgrins/spectrum/compare/issue264

Is it bonkers to merge branch issue264 into master now that it's 59 commits behind?

**Comment by bgrins** [comment]
OK, I've landed #405 on master which coordinates the events as discussed in #250.  Basically, a 'move' event fires when anything changes in the picker (including a palette click) and then a change only fires once the picker has been closed / submitted.   Flat pickers have sort of weird semantics since when the color is 'submitted' is unclear, so I'd suggest using the 'move' event in the case of a flat picker if you want the page to react to changes.

Here's a sample of the original fiddle using the latest master: http://jsfiddle.net/qfo2ykt0/7/.  I think this resolves the issue but let me know if not.
